server/src/main/java/org/opensearch/index/translog/RemoteFsTranslog.java

```diff
@@ -425,18 +425,18 @@ public void trimUnreferencedReaders() throws IOException {
         // cleans up remote translog files not referenced in latest uploaded metadata.
         // This enables us to restore translog from the metadata in case of failover or relocation.
         Set<Long> generationsToDelete = new HashSet<>();
-        for (long generation = minRemoteGenReferenced - 1; generation >= 0; generation--) {
+        for (long generation = minRemoteGenReferenced - 1 - indexSettings().getRemoteTranslogExtraKeep(); generation >= 0; generation--) {
             if (fileTransferTracker.uploaded(Translog.getFilename(generation)) == false) {
                 break;
             }
             generationsToDelete.add(generation);
         }
+        translogTransferManager.deleteStaleTranslogMetadataFilesAsync(remoteGenerationDeletionPermits::release);
         if (generationsToDelete.isEmpty() == false) {
             deleteRemoteGeneration(generationsToDelete);
-            translogTransferManager.deleteStaleTranslogMetadataFilesAsync(remoteGenerationDeletionPermits::release);
```

Comment:
Should we let this remain as is? The metdata deletion should remain in sync as well along with translog and ckp files. By this change, we would start making more remote calls which would cause more problems.

Reply:
I am on fence regarding this. But to make this change safe for now, let me do this as suggested .

Reply:
On every indexing request, this gets tirggers. So we do need this change.